[spmd compile api] pre-flatten state container and pass the flattened state container to transforms

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98233
* #98232
* #98231
* #98230
* #98229

Move the responsibility of flattening the input arguments from the graph module to the caller. This serves two purposes:
- Transformations that add/remove state need to manipulate a state container that maintains the state tensors in the same order as they appear in graph placeholders.
- Reduced runtime cost. The state container is only flattened once upfront.

Differential Revision: [D44637175](https://our.internmc.facebook.com/intern/diff/D44637175/)

Commit:
[spmd compile api] pre-flatten state container and pass the flattened state container to transforms

Move the responsibility of flattening the input arguments from the graph module to the caller. This serves two purposes:
- Transformations that add/remove state need to manipulate a state container that maintains the state tensors in the same order as they appear in graph placeholders.
- Reduced runtime cost. The state container is only flattened once upfront.

Differential Revision: [D44637175](https://our.internmc.facebook.com/intern/diff/D44637175/)

[ghstack-poisoned]

yifuwang commented on torch/distributed/_spmd/api.py:
```diff
@@ -118,7 +119,7 @@ def replacement(self, orig_submodule: torch.nn.Module) -> torch.nn.Module:
     @abstractmethod
     def transform(
         self, gm: fx.GraphModule, schema_map: Dict[str, Schema]
-    ) -> fx.Graph:
```
Good catch! Fixed

Commit:
Update on "[spmd compile api] pre-flatten state container and pass the flattened state container to transforms"


Move the responsibility of flattening the input arguments from the graph module to the caller. This serves two purposes:
- Transformations that add/remove state need to manipulate a state container that maintains the state tensors in the same order as they appear in graph placeholders.
- Reduced runtime cost. The state container is only flattened once upfront.

Differential Revision: [D44637175](https://our.internmc.facebook.com/intern/diff/D44637175/)

[ghstack-poisoned]

Commit:
Update on "[spmd compile api] pre-flatten state container and pass the flattened state container to transforms"


Move the responsibility of flattening the input arguments from the graph module to the caller. This serves two purposes:
- Transformations that add/remove state need to manipulate a state container that maintains the state tensors in the same order as they appear in graph placeholders.
- Reduced runtime cost. The state container is only flattened once upfront.

Differential Revision: [D44637175](https://our.internmc.facebook.com/intern/diff/D44637175/)

[ghstack-poisoned]

mrshenli commented:
lint error looks real:

```
>>> Lint for torch/distributed/_spmd/api.py:

  Error (MYPY) [arg-type]
    Argument "orig_args" to "_PyTreeInfo" has incompatible type "None";
    expected "List[str]"

        170  |    gm._graph._codegen = _PyTreeCodeGenOutputsOnly(
        171  |        pytree_info=_PyTreeInfo(
        172  |            # pyre-ignore[6]
    >>> 173  |            orig_args=None,
        174  |            # pyre-ignore[6]
        175  |            in_spec=None,
        176  |            # pyre-ignore[16]

  Error (MYPY) [arg-type]
    Argument "in_spec" to "_PyTreeInfo" has incompatible type "None"; expected
    "TreeSpec"

        172  |            # pyre-ignore[6]
        173  |            orig_args=None,
        174  |            # pyre-ignore[6]
    >>> 175  |            in_spec=None,
        176  |            # pyre-ignore[16]
        177  |            out_spec=gm._graph._codegen.pytree_info.out_spec,
        178  |        )
```

Commit:
Update on "[spmd compile api] pre-flatten state container and pass the flattened state container to transforms"


Move the responsibility of flattening the input arguments from the graph module to the caller. This serves two purposes:
- Transformations that add/remove state need to manipulate a state container that maintains the state tensors in the same order as they appear in graph placeholders.
- Reduced runtime cost. The state container is only flattened once upfront.

Differential Revision: [D44637175](https://our.internmc.facebook.com/intern/diff/D44637175/)

[ghstack-poisoned]

Commit:
Update on "[spmd compile api] pre-flatten state container and pass the flattened state container to transforms"


Move the responsibility of flattening the input arguments from the graph module to the caller. This serves two purposes:
- Transformations that add/remove state need to manipulate a state container that maintains the state tensors in the same order as they appear in graph placeholders.
- Reduced runtime cost. The state container is only flattened once upfront.

Differential Revision: [D44637175](https://our.internmc.facebook.com/intern/diff/D44637175/)

[ghstack-poisoned]

Commit:
Update on "[spmd compile api] pre-flatten state container and pass the flattened state container to transforms"


Move the responsibility of flattening the input arguments from the graph module to the caller. This serves two purposes:
- Transformations that add/remove state need to manipulate a state container that maintains the state tensors in the same order as they appear in graph placeholders.
- Reduced runtime cost. The state container is only flattened once upfront.

Differential Revision: [D44637175](https://our.internmc.facebook.com/intern/diff/D44637175/)

[ghstack-poisoned]

yifuwang commented:
@yifuwang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44637175).

yifuwang commented:
@yifuwang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44637175).

Commit:
Update on "[spmd compile api] pre-flatten state container and pass the flattened state container to transforms"


Move the responsibility of flattening the input arguments from the graph module to the caller. This serves two purposes:
- Transformations that add/remove state need to manipulate a state container that maintains the state tensors in the same order as they appear in graph placeholders.
- Reduced runtime cost. The state container is only flattened once upfront.

Differential Revision: [D44637175](https://our.internmc.facebook.com/intern/diff/D44637175/)

[ghstack-poisoned]

yifuwang commented:
@yifuwang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44637175).

yifuwang commented:
https://github.com/pytorch/pytorch/pull/98392
